### DISCUSSION TITLE
Type alias circularly reference error

### DISCUSSION THREAD

**Discussion by luan-xiaokun** [eclipse-langium/langium#729] [Q&A] [answered]
Hi guys, I am using langium 0.5, and I got the error `Type alias 'Term' circularly references itself` when compiling some parse rules like
```
Term:
    Literal | BracketTerm;

BracketTerm:
    '(' Term ')';
```

If the `Term` rule is replaced with `Literal | '(' Term ')'` then there is no error. This somehow confuses me, am I missing something?

**Comment by msujew** [comment, marked as answer]
Hey @luan-xiaokun,

usually you would try to make `BracketTerm` infer the same type as `Term` for this. This prevents the circular type reference. See the following example in the arithmetics example:

https://github.com/langium/langium/blob/9f587815f9d14f433d22831ac11d3b9f11696d9f/examples/arithmetics/src/language-server/arithmetics.langium#L31-L34

**Reply by luan-xiaokun** [reply]
It works after adding `infers Term` to the `BracketTerm` rule, thanks for your help!
Actually, I want the `Term` rule to also include tuple term, so I try to eliminate the left recursion by using
```
BracketTerm infers Term:
    '(' Term ({infer TupleTerm.members+=current} (',' members+=Term)) ')';
```
I have a question for this parse rule, when I write `current`, does it refer to the first `Term` or the whole already seen part `'(' Term`?

**Reply by msujew** [reply]
There's no difference between that, since `current` only affects the AST, wherease `'('` is part of the CST.

**Reply by luan-xiaokun** [reply]
I see, thanks for your answering!
